src/responder.rs

```diff
@@ -1,33 +1,20 @@
 use std::convert::TryFrom;
-use std::future::Future;
-use std::marker::PhantomData;
-use std::pin::Pin;
-use std::task::{Context, Poll};
 
 use actix_http::error::InternalError;
 use actix_http::http::{
     header::IntoHeaderValue, Error as HttpError, HeaderMap, HeaderName, StatusCode,
 };
 use actix_http::{Error, Response, ResponseBuilder};
 use bytes::{Bytes, BytesMut};
-use futures_util::future::{err, ok, Either as EitherFuture, Ready};
-use futures_util::ready;
-use pin_project::pin_project;
 
 use crate::request::HttpRequest;
 
 /// Trait implemented by types that can be converted to a http response.
 ///
 /// Types that implement this trait can be used as the return type of a handler.
 pub trait Responder {
-    /// The associated error which can be returned.
-    type Error: Into<Error>;
-
-    /// The future response value.
-    type Future: Future<Output = Result<Response, Self::Error>>;
-
-    /// Convert itself to `AsyncResult` or `Error`.
-    fn respond_to(self, req: &HttpRequest) -> Self::Future;
+    /// Convert self to `Response`.
+    fn respond_to(self, req: &HttpRequest) -> Response;
```

Comment:
I think it would be better to have `-> Result<Response, impl Into<Error>>` so that pre-made responses from the `error` module can be used.

Reply:
After `respond_to` is called it's immediately matched by `HandlerService` and converted to `Response` anyway.
Return result here is kinda pointless by adding one additional branch and make the type larger by 8 bytes.
I believe a macro like this for error type would be a better alternative
```rust
macro_rules! err_respond {
    ($ty: ty) => (
        impl Responder for $ty {
            fn respond_to(self, _: &HttpRequest) -> Response {
                Response::from_error(self.into())
            }
        }
    )
}
```

Reply:
hard no on the macro approach tbh

Reply:
impl Responder for error type is mostly not useful anyway. as impl T would still be one type making a handler can return error and only error. It's part of the reason why not all crate error types impl Responder